packages/events/createResponderEventPlugin.js

```diff
@@ -0,0 +1,696 @@
+ * - `touchStart`              (`EventPluginHub` dispatches as usual)
+ * - `responderGrant/Reject`   (`EventPluginHub` dispatches as usual)
+ */
+
+type GlobalResponderHandlerType = {
+  onChange: (from: any, to: any, blockNativeResponder: any) => void,
+};
+
+type GlobalInteractionHandlerType = {
+  onChange: (numberActiveTouches: number) => void,
+};
+
+export default function createResponderEventPlugin(TopLevelTypes: {
+  topMouseDown: TopLevelType,
+  topMouseMove: TopLevelType,
+  topMouseUp: TopLevelType,
+  topScroll: TopLevelType,
+  topSelectionChange: TopLevelType,
+  topTouchCancel: TopLevelType,
+  topTouchEnd: TopLevelType,
+  topTouchMove: TopLevelType,
+  topTouchStart: TopLevelType,
+}) {
+  function isStartish(topLevelType) {
+    return (
+      topLevelType === TopLevelTypes.topMouseDown ||
+      topLevelType === TopLevelTypes.topTouchStart
+    );
+  }
+
+  function isMoveish(topLevelType) {
+    return (
+      topLevelType === TopLevelTypes.topMouseMove ||
+      topLevelType === TopLevelTypes.topTouchMove
+    );
+  }
+
+  function isEndish(topLevelType) {
+    return (
+      topLevelType === TopLevelTypes.topMouseUp ||
+      topLevelType === TopLevelTypes.topTouchEnd ||
+      topLevelType === TopLevelTypes.topTouchCancel
+    );
+  }
+
+  const ResponderTouchHistoryStore = createResponderTouchHistoryStore(
+    isStartish,
+    isMoveish,
+    isEndish,
+  );
+
+  /**
+   * Instance of element that should respond to touch/move types of interactions,
+   * as indicated explicitly by relevant callbacks.
+   */
+  let responderInst = null;
+
+  /**
+   * Count of current touches. A textInput should become responder iff the
+   * selection changes while there is a touch on the screen.
+   */
+  let trackedTouchCount = 0;
+
+  /**
+   * Last reported number of active touches.
+   */
+  let previousActiveTouches = 0;
+
+  const changeResponder = function(nextResponderInst, blockHostResponder) {
+    const oldResponderInst = responderInst;
+    responderInst = nextResponderInst;
+    if (ResponderEventPlugin.GlobalResponderHandler !== null) {
+      ResponderEventPlugin.GlobalResponderHandler.onChange(
+        oldResponderInst,
+        nextResponderInst,
+        blockHostResponder,
+      );
+    }
+  };
+
+  const startDependencies = [
+    TopLevelTypes.topTouchStart,
+    TopLevelTypes.topMouseDown,
+  ];
+  const moveDependencies = [
+    TopLevelTypes.topTouchMove,
+    TopLevelTypes.topMouseMove,
+  ];
+  const endDependencies = [
+    TopLevelTypes.topTouchCancel,
+    TopLevelTypes.topTouchEnd,
+    TopLevelTypes.topMouseUp,
+  ];
+
+  const eventTypes = {
+    /**
+     * On a `touchStart`/`mouseDown`, is it desired that this element become the
+     * responder?
+     */
+    startShouldSetResponder: {
+      phasedRegistrationNames: {
+        bubbled: 'onStartShouldSetResponder',
+        captured: 'onStartShouldSetResponderCapture',
+      },
+      dependencies: startDependencies,
```

Comment:
The only thing I really changed here besides wrapping it in a factory function and adding type annotations was that I added dependencies to all event types (and thus upstreaming the changes required by RNW).

RN ignores those but the Flow type system still requires them.

Reply:
Will this be a breaking change for RNW? Technically we treat unstable* things as semver-exempt but it would be nice if it "just worked" for existing RNW users.

If not, it's okay, but would be nice to have a PR to RNW ready that shows how to prepare for this change.

Reply:
Yeah that will be a breaking change. The current version of RNW will overwrite the `dependencies` with the invalid top level strings and will patch the ResponderEventPlugin with an additional guard that won't work anymore. Here's the PR: https://github.com/necolas/react-native-web/pull/908

I _could_ build a PR to RNW that works on both the current version (using `top*` strings) as well as the next version (which hopefully uses my work) to smooth migration.